### DISCUSSION TITLE
ECDH-ES+A256KW is not supported as a valid `alg` when fetching a key from jwks

### DISCUSSION THREAD

**Discussion by xiazihang** [panva/jose#601] [Q&A] [answered]
Hi team, 
Recently Im working on a feature which need to implement the jose, and we'll need to fetch the public keys from a jwks endpoint, however, one of our public key was generated based on `ECDH-ES+A256KW` alg, after I read the source code of `jose` lib, at the line 7 in /node_modules/jose/dist/node/cjs/jwks/local.js. The method getKtyFromAlg(), it only support `RS`, `PS`, `ES`, `Ed`, other `alg` will throw error. which cannot work in our case.
![image](https://github.com/panva/jose/assets/19681630/2ab89c90-70ef-4143-91a5-996ab6f6bcbb)

I'd like to ask if you can support this? and also curios why do you design the search key logic like this? why can't we pass in the `kty` as one of the parameter in `protectedHeader` type?
Thanks

**Comment by panva** [comment]
Hi @xiazihang 

The [createRemoteJWKSet](https://github.com/panva/jose/blob/v5.0.1/docs/functions/jwks_remote.createRemoteJWKSet.md) is not intended for use with JWE. I thought that was clear from its documentation or typings, which mention only JWS, not JWE, but clearly I was wrong there and will try to improve.

The function serves to resolve public keys only because key resolver functions for JWS/JWT need to return a public/secret key. On the other hand, JWE consuming function key resolvers need to resolve a private key to decrypt, which is never distributed via a public JWKS endpoint.

In short, the function is meant for JWS, not JWE and that's why the "search key logic", as you call it, is the way it is.

**Reply by xiazihang** [reply]
Hi @panva,
Thanks very much for you quick reply and I think i get your point here, but still would like to share some of my thoughts.
As you mention that this particular method is intended for use with JWS only, that's explain why the method only look for `use=sig`, but in the real world cases, ppl might put the both public keys of `enc` and `sig` use together in the JWKS endpoint (our case is! ^ ^), and i think it's also make sense to make jwks as a usage-ignostic tool as JWS and JWE are come together in some cases.

Besides:

> JWE consuming function key resolvers need to resolve a private key to decrypt

This is only for decrypting a JWE, however ppl might need to generate a JWE which requires a `enc` public key and also could be coming from a JWKS endpoint.

Maybe can also take this into consideration to see if we can improve the logic a bit here.

Thanks again!

**Reply by panva** [reply]
> that's explain why the method only look for use=sig

Actually, the method looks for keys that either do not have a `use` at all, or ones where the `use` is `sig`. The JWKS endpoint may very well have keys that are sig, enc, or none of those two and will still work to resolve the proper public key used for JWS verification (given the JSON Web Key Set is well formed).

> however ppl might need to generate a JWE which requires a enc public key and also could be coming from a JWKS endpoint.

I get that, but this module's functions that encrypt do not accept a resolver function. What I get from your comments is that you wanted to use the method like so `await jose.createRemoteJWKSet(url)({ alg: 'ECDH-ES+A256KW' })` which is simply not its intended use, the intended use is well shown in the examples in the function's documentation. I intend to support the 90~ish% use-case for JWKS which is signature verification, not JWE encryption.

What you want is not hard to do without this function, simple fetch, then search based on your particular needs and subsequent `jose.importJWK` will do.

**Reply by xiazihang** [reply]
Ah if that is the case, could you kindly point me out how to do it without this function? I promise i've read your README haha and this is where it led me to.
So my use case is:
1. Need to download the public keys for both `enc` and `sig` provided by the client side
2. Use my own sig private key to generate a JWS -> jws
3. Find the enc public key from the JWKSet to generate a JWE -> jwe

what i've done with `jose` now is:
```
const keyStore = jose.createRemoteJWKSet(
        new URL(jwksEndpoint),
    );
const encPublicKey = await keyStore({
        alg: 'ECDH-ES+A256KW',
        kid: 'keyId',
    });
const jws = await new jose.CompactSign(
        textEncoder.encode(JSON.stringify(request)),
    )
        .setProtectedHeader({ kid: kid, alg: alg })
        .sign(sigPrivateKey); // my own sig private

const jwe = await new jose.CompactEncrypt(textEncoder.encode(jws))
        .setProtectedHeader({ alg: 'ECDH-ES+A256KW', enc: 'A256CBC-HS512' })
        .encrypt(encPublicKey); // from JWKS which can not be found
```
What else functions that i can use to achieve the same?

**Reply by panva** [reply, marked as answer]
As I wrote, use [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) to get the jwks, parse it, find the JWK you want to use, and then use [`jose.importJWK`](https://github.com/panva/jose/blob/v5.x/docs/functions/key_import.importJWK.md).

**Reply by xiazihang** [reply]
Well that's what I try to avoid to do manually as your createRemoteJWKSet() already implement cache, coolDown etc. 
but anyway, i get your point. thanks again!
